src/hashset.h

```diff
@@ -0,0 +1,188 @@
+ * implementation and documentation of functions, see comments in hashset.c.
+ *
+ * The elements in a hashset are of a user-defined type, but an element needs to
+ * contain a key. It can represent a key-value entry, or it can be just a key,
+ * if set semantics are desired.
+ *
+ * Terminology:
+ *
+ * hashset
+ *         An instance of the data structure.
+ *
+ * element
+ *         An element in the hashset. This may be of the same type as the key,
+ *         or a struct containing a key and other fields.
+ * key
+ *         The part of the element used for looking the element up in the hashset.
+ *         May be the entire element or a struct field within the element.
+ *
+ * type
+ *         A struct containing callbacks, such as hash function, key comparison
+ *         function and how to get the key in an element.
+ */
+
+#include "fmacros.h"
+#include <stddef.h>
+#include <stdint.h>
+#include <unistd.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+typedef struct hashsetBucket hashsetBucket;
+typedef struct hashsetStats hashsetStats;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
+    /* Hash function. Defaults to hashing the bits in the pointer, effectively
+     * treating the pointer as an integer. */
+    uint64_t (*hashFunction)(const void *key);
+    /* Compare function, returns 0 if the keys are equal. Defaults to just
+     * comparing the pointers for equality. */
+    int (*keyCompare)(hashset *s, const void *key1, const void *key2);
+    /* Callback to free an element when it's overwritten or deleted.
+     * Optional. */
+    void (*elementDestructor)(hashset *s, void *element);
+    /* Callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. Both tables are already created. */
+    void (*rehashingStarted)(hashset *s);
+    /* Invoked at the end of rehashing. Both tables still exist and are cleaned
+     * up after this callback. */
+    void (*rehashingCompleted)(hashset *s);
+    /* Track memory usage using this callback. It is called with a positive
+     * number when the hashset allocates some memory and with a negative number
+     * when freeing. */
+    void (*trackMemUsage)(hashset *s, ssize_t delta);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
```

Comment:
Instant = not incremental.

It's used for the command table lookup which is accessed by I/O threads in parallel. Command lookup is offloaded to I/O threads since Valkey 8. It's not thread safe to modify it so we make sure it doesn't change while the I/O threads are running. The command table only changes during MODULE LOAD and MODULE UNLOAD, so the I/O threads are stopped during the execution of these commands, and of course incremental rehashing can't be running either while the I/O threads access it.